src/metaopt/algo/base.py

```diff
@@ -0,0 +1,186 @@
+           hyperparameter names to values.
+
+        """
+        log.debug("Creating Algorithm object of %s type with parameters:\n%s",
+                  type(self).__name__, hypers)
+        self._space = space
+        # Instantiate tunable parameters of an algorithm
+        for varname, hyper in six.iteritems(hypers):
+            # Check if tunable element is another algorithm
+            if isinstance(hyper, dict) and len(hyper) == 1:
+                subalgo_type = list(hyper)[0]
+                subalgo_hypers = hyper[subalgo_type]
+                if isinstance(subalgo_hypers, dict):
+                    hyper = OptimizationAlgorithm(subalgo_type,
+                                                  space, **subalgo_hypers)
+
+            setattr(self, varname, hyper)
+
+    @abstractmethod
+    def suggest(self, num=1):
+        """Suggest a `num`ber of new sets of parameters.
+
+        :param num: how many sets to be suggested.
+
+        .. note:: New parameters must be compliant with the problem's domain
+           `metaopt.algo.space.Space`.
+        """
+        pass
+
+    @abstractmethod
+    def observe(self, points, results):
+        """Observe evaluation `results` corresponding to list of `points` in
+        space.
+
+        Parameters
+        ----------
+        points : list of tuples of array-likes
+           Points from a `metaopt.algo.space.Space`.
+           Evaluated problem parameters by a consumer.
+        results : list of dicts
+           Contains the result of an evaluation; partial information about the
+           black-box function at each point in `params`.
+
+        Result
+        ------
+        objective : numeric
+           Evaluation of this problem's objective function.
+        gradient : 1D array-like, optional
+           Contains values of the derivatives of the `objective` function
+           with respect to `params`.
+        constraint : list of numeric, optional
+           List of constraints expression evaluation which must be greater
+           or equal to zero by the problem's definition.
+
+        """
+        pass
+
+    @property
+    def is_done(self):
+        """Return True, if an algorithm holds that there can be no further improvement."""
+        return False
+
+    def score(self, point):
+        """Allow algorithm to evaluate `point` based on a prediction about
+        this parameter set's performance. Return a subjective measure of expected
+        performance.
+
+        By default, return the same score any parameter (no preference).
+        """
+        return 0
+
+    def judge(self, point, measurements):
+        """Inform an algorithm about online `measurements` of a running trial.
+
+        The algorithm can return a dictionary of data which will be provided
+        as a response to the running environment. Default is None response.
+
+        """
+        return None
+
+    @property
+    def should_suspend(self):
```

Comment:
It is a method to be used for polling. In my mind, worker shall first call `judge` to effectively update internal state of an `OptimizationAlgorithm` and then when needed `should_suspend` shall provide with algo's opinion on if worker should suspend.

I did not merge it with method `judge` because I mean `judge` to be a server callback in an online server(worker) - client(user's script) communication. Useful for reinforcement learning implementation or exotic online suggestion stuff which I do not know (but may be interesting research-wise).

Reply:
ok then the description of judge is not clear enough. It seems like a method for evaluating and returning something, it does sound like the algorithm is going to its internal state while judging.